### ISSUE TITLE
Understanding the CUDNN_STATUS_NOT_SUPPORTED warning

### ISSUE THREAD

**Issue by Kaspso** [ultralytics/ultralytics#11781]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

Hey, I am performing keypoint detection on airplanes using the yolov8n pose network. I have annotated 10 thousand images with keypoints and have been training the model upon them. But I have noticed this weird warning I get whenever I try to run the training:

**C:\Users\kaspe\Anaconda3\envs\YOLOEnv\Lib\site-packages\torch\nn\modules\conv.py:456: UserWarning: Plan failed with a cudnnException: CUDNN_BACKEND_EXECUTION_PLAN_DESCRIPTOR: cudnnFinalize Descriptor Failed cudnn_status: CUDNN_STATUS_NOT_SUPPORTED (Triggered internally at C:\cb\pytorch_1000000000000\work\aten\src\ATen\native\cudnn\Conv_v8.cpp:919.)
  return F.conv2d(input, weight, bias, self.stride,**

I have seen other mention that it might be too big of a batch size in my training data, but even very small batch sizes does not seem to make it go away. I am using Pytorch 2.3.0 with Cuda 11.8 and cudnn 8.7.0 are all supposed to function together so this does not explain it either.

Here is my full output when training:

(YOLOEnv) PS C:\Users\kaspe\YOLOV8Project> & C:/Users/kaspe/Anaconda3/envs/YOLOEnv/python.exe c:/Users/kaspe/YOLOV8Project/yolokeypointdetection.py
Ultralytics YOLOv8.2.11 🚀 Python-3.11.9 torch-2.3.0 CUDA:0 (NVIDIA T500, 4096MiB)
engine\trainer: task=pose, mode=train, model=yolov8n-pose.pt, data=C:/Users/kaspe/YOLOV8Project/LucasDataPlus/data.yaml, epochs=70, time=None, patience=100, batch=16, imgsz=512, save=True, save_period=-1, cache=False, device=None, workers=8, project=None, name=train17, exist_ok=False, pretrained=True, optimizer=auto, verbose=True, seed=0, deterministic=True, single_cls=False, rect=False, cos_lr=False, close_mosaic=10, resume=False, amp=True, fraction=1.0, profile=False, freeze=None, multi_scale=False, overlap_mask=True, mask_ratio=4, dropout=0.0, val=True, split=val, save_json=False, save_hybrid=False, conf=None, iou=0.7, max_det=300, half=False, dnn=False, plots=True, source=None, vid_stride=1, stream_buffer=False, visualize=False, augment=False, agnostic_nms=False, classes=None, retina_masks=False, embed=None, show=False, save_frames=False, save_txt=False, save_conf=False, save_crop=False, show_labels=True, show_conf=True, show_boxes=True, line_width=None, format=torchscript, keras=False, optimize=False, int8=False, dynamic=False, simplify=False, opset=None, workspace=4, nms=False, lr0=0.01, lrf=0.01, momentum=0.937, weight_decay=0.0005, warmup_epochs=3.0, warmup_momentum=0.8, warmup_bias_lr=0.1, box=7.5, cls=0.5, dfl=1.5, pose=12.0, kobj=1.0, label_smoothing=0.0, nbs=64, hsv_h=0.015, hsv_s=0.7, hsv_v=0.4, degrees=0.0, translate=0.1, scale=0.5, shear=0.0, perspective=0.0, flipud=0.0, fliplr=0.5, bgr=0.0, mosaic=1.0, mixup=0.0, copy_paste=0.0, auto_augment=randaugment, erasing=0.4, crop_fraction=1.0, cfg=None, tracker=botsort.yaml, save_dir=runs\pose\train17
Overriding model.yaml kpt_shape=[17, 3] with kpt_shape=[22, 2]

                   from  n    params  module                                       arguments
  0                  -1  1       464  ultralytics.nn.modules.conv.Conv             [3, 16, 3, 2]
  1                  -1  1      4672  ultralytics.nn.modules.conv.Conv             [16, 32, 3, 2]
  2                  -1  1      7360  ultralytics.nn.modules.block.C2f             [32, 32, 1, True]
  3                  -1  1     18560  ultralytics.nn.modules.conv.Conv             [32, 64, 3, 2]
  4                  -1  2     49664  ultralytics.nn.modules.block.C2f             [64, 64, 2, True]
  5                  -1  1     73984  ultralytics.nn.modules.conv.Conv             [64, 128, 3, 2]
  6                  -1  2    197632  ultralytics.nn.modules.block.C2f             [128, 128, 2, True]
  7                  -1  1    295424  ultralytics.nn.modules.conv.Conv             [128, 256, 3, 2]
  8                  -1  1    460288  ultralytics.nn.modules.block.C2f             [256, 256, 1, True]
  9                  -1  1    164608  ultralytics.nn.modules.block.SPPF            [256, 256, 5]
 10                  -1  1         0  torch.nn.modules.upsampling.Upsample         [None, 2, 'nearest']
 11             [-1, 6]  1         0  ultralytics.nn.modules.conv.Concat           [1]
 12                  -1  1    148224  ultralytics.nn.modules.block.C2f             [384, 128, 1]
 13                  -1  1         0  torch.nn.modules.upsampling.Upsample         [None, 2, 'nearest']
 14             [-1, 4]  1         0  ultralytics.nn.modules.conv.Concat           [1]
 15                  -1  1     37248  ultralytics.nn.modules.block.C2f             [192, 64, 1]
 16                  -1  1     36992  ultralytics.nn.modules.conv.Conv             [64, 64, 3, 2]
 17            [-1, 12]  1         0  ultralytics.nn.modules.conv.Concat           [1]
 18                  -1  1    123648  ultralytics.nn.modules.block.C2f             [192, 128, 1]
 19                  -1  1    147712  ultralytics.nn.modules.conv.Conv             [128, 128, 3, 2]
 20             [-1, 9]  1         0  ultralytics.nn.modules.conv.Concat           [1]
 21                  -1  1    493056  ultralytics.nn.modules.block.C2f             [384, 256, 1]
 22        [15, 18, 21]  1    987655  ultralytics.nn.modules.head.Pose             [1, [22, 2], [64, 128, 256]]  
YOLOv8n-pose summary: 250 layers, 3247191 parameters, 3247175 gradients, 9.1 GFLOPs

Transferred 361/397 items from pretrained weights
Freezing layer 'model.22.dfl.conv.weight'
AMP: running Automatic Mixed Precision (AMP) checks with YOLOv8n...
C:\Users\kaspe\Anaconda3\envs\YOLOEnv\Lib\site-packages\torch\nn\modules\conv.py:456: UserWarning: Plan failed with a cudnnException: CUDNN_BACKEND_EXECUTION_PLAN_DESCRIPTOR: cudnnFinalize Descriptor Failed cudnn_status: CUDNN_STATUS_NOT_SUPPORTED (Triggered internally at C:\cb\pytorch_1000000000000\work\aten\src\ATen\native\cudnn\Conv_v8.cpp:919.)
  return F.conv2d(input, weight, bias, self.stride,
AMP: checks passed ✅

**Comment by glenn-jocher** [comment]
Hello! It seems you're encountering some issues with cuDNN when using YOLOv8 for keypoint detection. The `CUDNN_STATUS_NOT_SUPPORTED` warning sometimes indicates a compatibility issue between your configuration and cuDNN's capabilities, particularly with specific convolution operations that your model might be using.

Even though your software versions (PyTorch, CUDA, and cuDNN) are compatible with each other, the error might be due to specific layer parameters or operations in the model that are not supported by your GPU. Here’s what you could try:
- **Verify Compatibility**: Ensure that all layer configurations (like kernel sizes, padding, etc.) within your model are supported by cuDNN on your particular GPU.
- **Update or Roll-back Software Versions**: Although your current versions should work together, sometimes slight differences in update patches might cause unexpected issues. Trying a slightly different combination may help.
- **Adjust Model Architecture**: It’s also possible that adjusting specific layer parameters might avoid this issue.
  
If none of these help, please provide more details about your model architecture and any modifications you’ve made to it. This will help in diagnosing the issue better! 😊

**Comment by gerrylwk** [comment]
I encountered a similar error when trying to run YOLOv8 from the official repository:
`
from ultralytics import YOLO
import cv2
import torch
# torch.backends.cudnn.benchmark = True
# torch.backends.cudnn.enabled=False
# torch.backends.cudnn.deterministic = True
# Load a model
model = YOLO("yolov5su.pt")  # load a pretrained model (recommended for training)

im2 = cv2.imread("/home/smuna/MPCount/inputs/Screenshot from 2024-05-16 14-59-05.png")
results = model.predict(source=im2, conf=0.1, half=True, classes=[0])
`

The fix for me was to run it on benchmark mode OR disabling cudnn. Hope this helps! Credits to ptrblack from official Torch forums

**Comment by glenn-jocher** [comment]
@gerrylwk hey there! Thanks for sharing your solution! 🌟 It's great to see community members helping each other out. Using `torch.backends.cudnn.benchmark = True` can indeed speed up the runtime by allowing cuDNN to automatically find the most efficient algorithms for your specific configuration. Disabling cuDNN can be a useful fallback when encountering compatibility issues. Both approaches are valid depending on the specific needs and setup. Keep up the great work, and happy coding with YOLOv8! 😊

**Comment by tripathi-genius** [comment]
@gerrylwk thanks for the solution. I faced a similar issue, but it was solved when I used
`from ultralytics import YOLO
import cv2
import torch
torch.backends.cudnn.benchmark = True
# torch.backends.cudnn.enabled=False
torch.backends.cudnn.deterministic = True
# Load a model
model = YOLO("yolov5su.pt")  # load a pretrained model (recommended for training)

im2 = cv2.imread("test.png")
results = model.predict(source=im2, conf=0.1, half=True, classes=[0])`

It was on NVIDIA-RTX 3050 GPU. It seems to be a viable solution.

**Comment by aysent911** [comment]
I added only the lines :
```
torch.backends.cudnn.benchmark = True
torch.backends.cudnn.deterministic = True
```
and the warning disappeared. I was tuning resnet50 on NVIDIA-RTX 3050

**Comment by glenn-jocher** [comment]
Hello @aysent911,

Thank you for sharing your experience! It's great to hear that enabling `torch.backends.cudnn.benchmark` and setting `torch.backends.cudnn.deterministic` resolved the warning for you. These settings can indeed help optimize performance and ensure reproducibility:

- `torch.backends.cudnn.benchmark = True` allows cuDNN to find the most efficient algorithms for your hardware, which can speed up your model training and inference.
- `torch.backends.cudnn.deterministic = True` ensures that operations are deterministic, which is useful for reproducibility.

Here's a quick summary of your solution for others who might encounter a similar issue:

```python
import torch
from ultralytics import YOLO
import cv2

# Enable cuDNN benchmark and deterministic mode
torch.backends.cudnn.benchmark = True
torch.backends.cudnn.deterministic = True

# Load a pretrained YOLO model
model = YOLO("yolov5su.pt")

# Read an image
im2 = cv2.imread("test.png")

# Run prediction
results = model.predict(source=im2, conf=0.1, half=True, classes=[0])
```

This approach can be particularly helpful when working with models like ResNet50 on GPUs like the NVIDIA RTX 3050. If anyone else is facing similar issues, give this a try! 😊

Thanks again for contributing to the community, and happy coding!
